enhancements/update/targeted-update-edge-blocking.md

```diff
@@ -0,0 +1,400 @@
+* `clusters` (optional, [object][json-object]), defining the subset of affected clusters.
+  If any `clusters` property matches a given cluster, the edge should be blocked for that cluster.
+  If `clusters` is unset, the edge is blocked for all clusters.
+  * `promql` (optional, [string][json-string]), with a [PromQL][] query describing affected clusters.
+    This query will be evaluated on the local cluster, so it has access to data beyond the subset that is [uploaded to Telemetry][uploaded-telemetry].
+    The query should return a 0 if the update should be allowed and a 1 if the update should be blocked.
+  * FIXME: [`platform` or other properties that are easier to evaluate than PromQL?](#non-promql-filters)
+
+[The schema version][graph-data-schema-version] would also be bumped to 1.1.0, because this is a backwards-compatible change.
+Consumers who only understand graph-data schema 1.0.0 would ignore the `clusters` property and block the edge for all clusters.
+The alternative of failing open is discussed [here](#failing-open).
+
+### Enhanced Cincinnati JSON representation
+
+[The Cincinnati graph API][cincinnati-api] will be extended with a new top-level `conditionalEdges` property, with an array of conditional edge [objects][json-object] using the following schema:
+
+* `from` (required, [string][json-string]), with the `version` of the starting node.
+* `to` (required, [string][json-string]), with the `version` of the ending node.
+* `qualifiers` (optional, [array][json-array], with qualifications around the recommendation.
+  Each entry is an [object][json-object] with the following schema:
+  * `url` (optional, [string][json-string]), with a URI documenting the issue, as described in [the blocked-edges section](#enhanced-graph-data-schema-for-blocking-edges).
+  * FIXME: [also reason/message?](#metadata-to-include-when-blocking-a-conditional-request)
+  * `clusters` (optional, [object][json-object]).
+    If any `clusters` property matches a given cluster, the edge is not recommended for that cluster, as described in [the blocked-edges section](#enhanced-graph-data-schema-for-blocking-edges).
+    If `clusters` is unset, the edge is blocked for all clusters.
+    * `promql` (optional, [object][json-object]), defining the subset of affected clusters, as described in [the blocked-edges section](#enhanced-graph-data-schema-for-blocking-edges).
+    * FIXME: [`platform` or other properties that are easier to evaluate than PromQL?](#non-promql-filters)
+
+### Enhanced ClusterVersion representation
+
+[The ClusterVersion `status`][api-cluster-version-status] will be extended with a new `conditionalUpdates` property:
+
+```go
+// conditionalUpdates contains the list of updates that are
+// conditionally recommended for this cluster. Consumers interested
+// in the set of updates that are actually recommended for this
+// cluster should use availableUpdates. This list may be empty if no
```

Comment:
```
// Consumers interested
// in the set of updates that are actually recommended for this
// cluster should use availableUpdates.
```
The relationship between and `conditionalUpdates` and `availableUpdates` is confusing. We should keep it simple i.e. users needs to see only in `availableUpdates` and if there is nothing in available updates then they should find the rule that has eliminated the versions from it.

Reply:
Can we rename `conditionalUpdates` to `notRecommendedByOSUS` or something similar?

Reply:
> We should keep it simple i.e. users needs to see only in availableUpdates...

That's what this comment is saying, isn't it?  And then folks who are interested in the new-with-this-enhancement supported-but-not-recommended-for-this-cluster edges look at `conditionalUpdates` entries with `Recommended!=True`.  And between the two, that's all the supported edges.

Reply:
the comment says:
```
// conditionalUpdates contains the list of updates that are
// conditionally recommended for this cluster.
```

If entries in conditionalUpdates can contain a "Recommended==false" field/attribute, then i'd say the comment isn't accurate.  It's not the list of edges that are `conditionally recommended`.  It's the list of edges which have conditions on them, and your cluster may or may not meet the condition that allows/recommends the upgrade.

If your cluster does meet the requirements such that a conditionalEdge upgrade should be allowed, will that update also appear in availableUpdates?

Reply:
```
// availableUpdates contains the subset of conditionalUpdates that
// apply to this cluster.  Updates which appear in conditionalUpdates
// but not in availableUpdates may expose this cluster to known
// issues.
```

answers some of my questions, but i think still calls for cleanup of the godoc on conditionalUpdates.  These are not updates that are "conditionally recommended for this cluster", they are "updates that have conditions on them, your cluster may or may not meet the conditions that would allow this upgrade to be recommended for your cluster"

i also think we need to standardize on how we're using the term condition.  If the cluster "meets the condition" does that mean the edge is blocked (we detected the vulnerable condition on your cluster and blocked the edge for you), or does it mean the edge is allowed (you fulfill the conditions required to allow this edge)?

Reply:
I found this confusing as well and my opinion of what `conditionalUpdates` contained changed once I read the `availbleUpdates` comment.